File: Python/confidence_pedigree.py
# Load libraries
import plotly.graph_objects as go
import pandas as pd
import plotnine as plt
# Load  Dataframe
data = {"EMD": [3,3,3,5,3,2,3,5],
        "P&D": [2,4,3,1,2,2,3,5],
        "O&S": [2,2,"",1,2,2,3,5],
        "Comments": ["1", "2", "3", "4", "5", "6", "7", "Key performance requirements well-defined"]}

# create dataframe
df = pd.DataFrame(data, columns = ["EMD", "P&D", "O&S", "Comments"])
# transpose dataframe
df = df.T

# create axis labels
Phase = ["EMD", "P&D", "O&S"] 
Confidence_Enablers = ["Overall Assessment  ", "Budget Equals Estimate  ", "Risk Assessment  <br>(Cost Schedule Technical)  ", 
                                "Crosschecks  ", "Cost & Methodology  ", "Schedule Baseline  ", 
                                "Engineering Technical Baseline  ", "Requirements Definition  "]

# create heatmap formatting

font = go.layout.Font(family = "Times New Roman",
                  size = 10,
                  color = 'black')
title = go.layout.Title(text = "<i>Program 'Confidence' Pedigree<i>",
                        font = {"size": 18})
xaxistitle = go.layout.xaxis.Title(text = "<b>Program Phase</b>",
                                  font = {"size": 16})
xaxis = go.layout.XAxis(side = "top",
                        title = xaxistitle,
                        mirror = True,
                        linewidth = 2,
                        linecolor = "black",
                        showline = True)
yaxistitle = go.layout.yaxis.Title(text = "<b>Confidence Enablers<b>",
                                  font = {"size": 16})                        
yaxis = go.layout.YAxis(title = yaxistitle,
                        mirror = True,
                        linewidth = 2,
                        linecolor = "black",
                        showline = True)
layout = go.Layout(font = font,
                   xaxis = xaxis,
                   yaxis = yaxis,
                   plot_bgcolor = "white",
                   title = title)

# make dataframe usable by heatmap (create array)
def get_zvals_as_array(x):
  df2 = x.iloc[0:3]
  result = []
  for (columnName, columnData) in df2.iteritems():
    column = columnData.values
    result.append(column)
  return result

zvalues = get_zvals_as_array(df)

def get_comments_as_array(x):
  df2 = x.iloc[-1]
  result = []
  for (columnName, columnData) in df2.iteritems():
    column = []
    column1 = columnData
    column.append(column1)
    column.append(column1)
    column.append(column1)
    result.append(column)
  return result


textvalues = get_comments_as_array(df)
# create heatmap

fig = go.Figure(data = go.Heatmap(z = zvalues,
                                  x = Phase,
                                  y = Confidence_Enablers,
                                  hoverongaps = False,
                                  colorscale = "rdylgn",
                                  #showscale = False,
                                  hoverinfo = "text",
                                  text = textvalues),
                layout = layout)

# plot
fig.show()

File: R/confidence_pedigree.r
# Load libraries
library(plotly)
# Load Dataframe
df_confidence <- data.frame(confidence = c(3,3,3,5,3,2,3,5,
                                2,2,NA,2,2,2,3,5,
                                2,4,3,2,2,2,3,5),
                comments = c("1", "2", "3", 
            "4", "5", "6",
            "7", "8", "9",
            "10", "11", "12",
            "13", "14", "15",
            "16", "17", "18",
            "19", "20", "21",
            "22", "23", "24"))

# Create Matrix
m <- matrix(df_confidence[,1], nrow = 8, ncol = 3)
commentmatrix <- matrix(df_confidence[,2], nrow = 8, ncol = 3)

# plot matrix
t <- list(
  family = "Times New Roman",
  size = 16,
  color = 'black')

confidence_pedigree <- plot_ly(
    x = c("EMD", "O&S", "P&D"), 
    y = c("Overall Assessment", "Budget Equals Estimate", "Risk Assessment (Cost Schedule Technical)",
    "Crosschecks", "Cost & Methodology", "Schedule Baseline", "Engineering Technical Baseline", "Requirements Definition"),
    z = m,
    zmin = 1,
    zmax = 5, 
    colors = colorRamp(c("red", "yellow", "green")),
    #colors = list([1, "FF0000"], [2, "ff6600"], [3, "FFCC00"], [4, "99ff00"], [5, "33ff00"]),
    type = "heatmap",
    hoverinfo = "text",
    text = commentmatrix) %>% 
    layout(title = list(text = "<i>Confidence Pedigree<i>", x = .1),
          xaxis = list(side = "top", 
                      title = "Program Phase <br> <br>", 
                      mirror = TRUE,
                      linewidth = 2,
                      linecolor = "black",
                      showline = TRUE), 
          yaxis = list(categoryorder = "trace", 
                      title = "Confidence Enablers <br> <br> <br> <br> <br> <br> <br> <br>",
                      mirror = TRUE,
                      linewidth = 2,
                      linecolor = "black",
                      showline = TRUE), 
          margin = list(l=10, r=10, b=10, t=10),
          font = t) %>%
    colorbar(tickvals = c(1, 2, 3, 4, 5))

confidence_pedigree

